src/System.IO.Pipelines/src/System/IO/Pipelines/Pipe.cs

```diff
@@ -18,7 +20,14 @@ public sealed partial class Pipe
 
         private static readonly Action<object> s_signalReaderAwaitable = state => ((Pipe)state).ReaderCancellationRequested();
```

Comment:
This could be fixed with new scheduler to not go through pipe reference and maybe even not allocate

Reply:
For the cancellation token callback we can't do much. We pass this reader direction to Register.